Adding kubernetes module

erjohnso opened the pull request:
As discussed in https://github.com/ansible/ansible-modules-extras/issues/1139, here's the proposed module for Kubernetes resource management via the API.

As noted in a TODO (line 239), SSL certificate validation is effectively disabled.

gregdek commented:
Thanks @erjohnso for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes, which can be contributed by anyone other than the author of the new module:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

More information about the review process can be seen here: 
https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md

Thanks again for submitting your Ansible module!

sivel commented:
Here is some feedback from the ansible-testing project:

```
============================================================================
kubernetes.py
============================================================================
Traceback (most recent call last):
  File "ansible_testing/modules.py", line 351, in validate
    doc, examples, ret = get_docstring(self.path, verbose=True)
  File "/Users/matt/python_venvs/ansibledev/ansible/lib/ansible/utils/module_docs.py", line 74, in get_docstring
    doc = yaml.safe_load(child.value.s)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/__init__.py", line 93, in safe_load
    return load(stream, SafeLoader)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/__init__.py", line 71, in load
    return loader.get_single_data()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/constructor.py", line 37, in get_single_data
    node = self.get_single_node()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 55, in compose_document
    node = self.compose_node(None, None)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 110, in compose_sequence_node
    while not self.check_event(SequenceEndEvent):
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/parser.py", line 382, in parse_block_sequence_entry
    if self.check_token(BlockEntryToken):
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 116, in check_token
    self.fetch_more_tokens()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 220, in fetch_more_tokens
    return self.fetch_value()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 580, in fetch_value
    self.get_mark())
ScannerError: mapping values are not allowed here
  in "<string>", line 9, column 71:
     ... AML file. Currently, this module:
                                         ^

ERROR: Invalid or no DOCUMENTATION provided
WARNING: No RETURN provided
WARNING: Found Try/Except block without HAS_ assginment
WARNING: Redeclared basic.py variable or function: json
```

The `try/except` warning and the `json` warning are basically the same.  You already get this from `basic.py` so no need to import `json` or `simplejson` in your module at this time.

Looks like the big one is just that the `DOCUMENTATION` is malformed.

gregdek commented:
@erjohnso -- does this obsolete #1139?

Smana commented:
Thanks i'm going to test it for my playbooks : https://github.com/ansibl8s

Just a question please, what is the ansible version which includes this module ?

Smana commented:
I've a simple tasks as follows : 

```
- name: Look up pods
  kubernetes:
    api_endpoint: 10.115.99.1:8080
    url_username: admin
    url_password: changeme
    kubectl_get: pods
    state: kubectl
```

which returns a `null` value and i don't know why

```
TASK: [test_kube | Look up pods] ********************************************** 
<10.115.99.1> ESTABLISH CONNECTION FOR USER: root
<10.115.99.1> REMOTE_MODULE kubernetes url_username=admin kubectl_get=pods state=kubectl url_password=changeme api_endpoint=10.115.99.1:8080
<10.115.99.1> EXEC sshpass -d9 ssh -C -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/root/.ansible/cp/ansible-ssh-%h-%p-%r" -o Port=22 -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 10.115.99.1 /bin/sh -c 'LANG=C LC_CTYPE=C /usr/bin/python'
ok: [10.115.99.1] => {"api_response": null, "changed": false}
```

The api is callable via 

```
curl -u admin:toto http://10.115.99.1:8080/version
{
  "major": "1",
  "minor": "1",
  "gitVersion": "v1.1.2",
  "gitCommit": "3085895b8a70a3d985e9320a098e74f545546171",
  "gitTreeState": "clean"
}
```

Any idea ?

erjohnso commented:
@Smana - what kind of k8s cluster are your running (local linux, vagrant, gke, etc)?

What you have looks good (except for the different passwords `changeme` and `toto`). The only thing I see that's different than what I was using for development is that your endpoint is not using SSL. But since I'm using `fetch_url` from `ansible.module_utils.urls`, that may not be the issue.

What you have works for my k8s cluster running in GKE.

```
erjohnso@ansible-dev$ ansible-playbook -vvv pseudo-kubectl/getpods.yml
Using /home/erjohnso/ansible.cfg as config file
1 plays in pseudo-kubectl/getpods.yml

PLAY ***************************************************************************

TASK [Get pods] ****************************************************************
task path: /home/erjohnso/ansible-k8s-demos/pseudo-kubectl/getpods.yml:6
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o 'IdentityFile="/home/erjohnso/.ssh/google_compute_engine"' -o KbdInteractiveAuthentication=no -o 
PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/erjohnso/.ansible/cp/ansible-ss
h-%h-%p-%r -tt localhost (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1448544063.05-29836441762329`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1448544063.05-
29836441762329`")
<localhost> PUT /tmp/tmpnAYtMG TO /home/erjohnso/.ansible/tmp/ansible-tmp-1448544063.05-29836441762329/kubernetes
<localhost> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o 'IdentityFile="/home/erjohnso/.ssh/google_compute_engine"' -o KbdInteractiveAuthentication=no 
-o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/erjohnso/.ansible/cp/ansible
-ssh-%h-%p-%r [localhost]
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o 'IdentityFile="/home/erjohnso/.ssh/google_compute_engine"' -o KbdInteractiveAuthentication=no -o 
PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/erjohnso/.ansible/cp/ansible-ss
h-%h-%p-%r -tt localhost LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/erjohnso/.ansible/tmp/ansible-tmp-1448544063.05-29836441762329/kubern
etes; rm -rf "/home/erjohnso/.ansible/tmp/ansible-tmp-1448544063.05-29836441762329/" > /dev/null 2>&1
ok: [localhost] => {"api_response": {"apiVersion": "v1", "items": [], "kind": "PodList", "metadata": {"resourceVersion": "657825", "selfLink": "/api/v1/namespaces/default/pods"
}}, "changed": false, "invocation": {"module_args": {"api_endpoint": "123.45.67.89:443", "kubectl_get": "pods", "state": "kubectl", "url_password": "changeme", "url_u
sername": "admin"}, "module_name": "kubernetes"}}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

I don't have any pods running in my k8s cluster, but you can see that it's still hitting the k8s API and returning an empty `items` list.

I guess a few things to try would be to turn up ansible's verbosity and see if that shows anything new. But you may end up needing to add in some debugging lines in your local copy of ansibe / kubernetes module to look at the raw HTTP request to see if that shows what's wrong.

detiber commented:
@erjohnso @akbutcher and myself recently worked on prototyping the beginning of something similar to this, but we approached it a bit differently technically (mainly for supporting both Kubernetes and OpenShift with minimal duplication).

We built out the core functionality into module_utils and started building out multiple modules that exposed different kube concepts (namespace, pod, service, rc, etc)

I'm out of the office until the 16th, but I would like to see if there is a way to combine your approach with ours.

detiber commented:
@erjohnso below is the work I referenced above (keep in mind it is currently just a prototype we built in a day):

https://github.com/detiber/ansible/blob/hack_day/lib/ansible/module_utils/kubernetes.py
https://github.com/detiber/ansible-modules-core/tree/hack_day/cloud/kubernetes

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    "node": "/api/v1/nodes",
+    "persistentvolume": "/api/v1/persistentvolumes",
+    "persistentvolumeclaim":
+        "/api/v1/namespaces/{namespace}/persistentvolumeclaims",
```
This line should not have extra indentation.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    "pod": "/api/v1/namespaces/{namespace}/pods",
+    "podtemplate": "/api/v1/namespaces/{namespace}/podtemplates",
+    "replicationcontroller":
+        "/api/v1/namespaces/{namespace}/replicationcontrollers",
```
This line should not have extra indentation.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    choices: BOOLEANS
+
+requirements:
+    - "python >= 2.6"
```
There is no need for this requirement as Ansible itself requires it.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+
+requirements:
+    - "python >= 2.6"
+    - "PyYAML >= 3.10"
```
There is no need for this requirement as Ansible itself requires it.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+              "absent", "delete",
+              "update", "patch",
+              "replace", "put",
+              "kubectl"]
```
Three columns would look better ;o)

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+options:
+  api_endpoint:
+    description:
+      - the IPv4 API endpoint of the Kubernetes cluster
```
Make this a sentence (Capital letter at the beginning, period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    aliases: ["endpoint"]
+  inline_data:
+    description:
+      - the Kubernetes YAML data to send to the server endpoint
```
Make this a sentence (Capital letter at the beginning, period at the end).

jtyr replied:
The `endpoint` word could be highlighted by `I(endpoint)` to emphasise that it refers to the `api_endpoint` resp. `endpoint` option defined above.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    default: null
+  file_reference:
+    description:
+      - specify full path to a Kubernets YAML file to send to server endpoint
```
Make this a sentence (Capital letter at the beginning, period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    default: null
+  kubectl_api_versions:
+    description:
+      - Mimic the 'kubectl api-versions' command, values are ignored
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    default: null
+  kubectl_cluster_info:
+    description:
+      - Mimic the 'kubectl cluster-info' command, values are ignored
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    description:
+      - Mimic the 'kubectl get' command. Specify the object(s) to fetch such
+        as 'pods' or 'replicationcontrollers/mycontroller'. It does not
+        support shortcuts (e.g. 'po', 'rc', 'svc')
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    default: null
+  kubectl_namespace:
+    description:
+      - Specify the namespace to use for 'kubectl' commands
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    default: "default"
+  kubectl_version:
+    description:
+      - Mimic the 'kubectl version' command, values are ignored
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+  state:
+    description:
+      - The desired action to take on the Kubernetes data, or 'kubectl' to
+        mimic some kubectl commands
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+              "kubectl"]
+  url_password:
+    description:
+      - the HTTP Basic Auth password for the API endpoint
```
Make this a sentence (Capital letter at the beginning, period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    aliases: ["password"]
+  url_username:
+    description:
+      - the HTTP Basic Auth username for the API endpoint
```
Make this a sentence (Capital letter at the beginning, period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    description:
+      - enable/disable certificate validation. Note that this is set to false
+        until Ansible can support IP address based certificate hostname
+        matching (exists in >= python3.5.0)
```
Make this a sentence (period at the end).

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    aliases: ["username"]
+  validate_certs:
+    description:
+      - enable/disable certificate validation. Note that this is set to false
```
The word `false` should be marked as an value of the option by `C(false)`.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+# - Set 'required=true' for certificate_authority_data and ensure that
+#   ansible's SSLValidationHandler.get_ca_certs() can pick up this CA cert
+# - Set 'required=true' for the validate_certs param.
+def _decode_cert_data(module):
```
Does it make sense to mark this function as 'private' (underscore in front of the function name)? Even all the functions would be wrapped into a class, I would not make it 'private' either. The same allies for the function below.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+    return info, body
+
+
+def k8s_kubectl_get(module, url):
```
I would remove the `k8s_` from the name of this and all other functions.

jtyr commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,436 @@
+            changed, body = k8s_kubectl_get(module, kurl)
+            module.exit_json(changed=changed, api_response=body)
+        if kubectl_version:
+            _, body = _k8s_api_request(module, url + "/version")
```
I would replace the `_` variable name with `info` as you use above. The same applies to the bellow two conditionals.

jtyr commented:
I have done some basic code review. The code is also missing documentation of the return parameters which is required for Ansible v2.x. Look for examples [here](http://docs.ansible.com/ansible/developing_modules.html#example).

jtyr commented:
@detiber Your modules look interesting. Please mention me in the description once you create a PR as I'm keen to review it.

erjohnso commented:
@jtyr - thanks for the review and sorry for the slow response to feedback, but updating now.

jtyr commented:
Perhaps you could squash the commits into one to keep the final log clean and related only to the adding of the module.

erjohnso commented:
Ok, squashed and rebased to devel.

Looks like an independent community reviewer needs to add 'works_for_me' and 'passes_guidelines' before this will get merged.

@detiber - sorry I missed your comment earlier! So, we have two different approaches:
1. This pr allows users to re-use existing kubernetes yaml files (either in-line or references) and essentially provides an ansible native way to communicate with the k8s api server. A benefit here is that users can prototype their app directly in native k8s yaml files and then easily incorporate them into an ansible playbook.
2. Your approach allows users to model k8s resources in a more ansible idiomatic way, but of course requires the user to re-write any existing k8s yaml files to ansible playbooks. Not a big deal if there is no need to manage the k8s app files in two separate ways and they're only going to use ansible (versus in parallel with `kubectl`).

I had opted for the first approach to make k8s more accessible to ansible users wanting to try k8s in a more ansible-friendly way. There is a large body of [k8s examples](https://github.com/kubernetes/kubernetes/tree/master/examples) that can be easily wrapped in an ansible playbook. I've started on a demo that builds the guestbook example with ansible out of the raw k8s yaml files and it's trivial to setup.

However, I recognize that users will, after an initial getting started introduction, may want to model their app with more ansible native k8s representations.

Outside of some refactoring, I think the two can live together, perhaps even in the same module. wdyt?

detiber commented:
@erjohnso Indeed, I think both approaches can (and should) exist. To provide some context for my motivation:
In openshift-ansible (https://github.com/openshift/openshift-ansible) we are currently using the command module to execute the OpenShift client (which basically wraps and extends kubectl). To improve reliability, idempotency and allow for check mode to be useful, we'd want to build out resources in a composable way without having to template yaml or json files. Additionally, since OpenShift exposes both the Kubernetes and OpenShift apis, being able to extend the Kubernetes module(s) would go a long way to reducing duplication between OpenShift and Kubernetes modules. I'm probably being presumptuous here, but I would expect it could likely do the same for the GKE and Kubernetes module(s) as well.

I'm going to be working with a handful of people on Friday at 1:30 EST to draft a Design Doc that can bridge the two approaches. If you are interested (and available), I can forward you an invite to join.

erjohnso commented:
@detiber - sorry I couldn't make the make the meeting! How did it go? If you're open to it, please feel free to send me a copy of the doc to erjohnso at google dot com.

w.r.t. to this module and PR, should we scrap it in favor of something else, or get this merged and refactor?

bcoca commented:
As for the PR we might want to keep as is to have a 'experts/advanced' mode
in which you just repurpose existing json/yaml configs, but we also want
easy to use modules for users to update using module options instead of the
json/yaml.

Zempashi commented:
It seem's that url_username and url_password are not hooked up in the module ?! It's something that comes soon or that will need some contribution ?

erjohnso commented:
Hi @Zempashi - the `url_username` and `url_password` are used automatically by the `fetch_url()` method via [module params](https://github.com/erjohnso/ansible-modules-extras/blob/google-kubernetes/clustering/kubernetes.py#L273).  This method gets pulled from `ansible.module_utils.urls`.

erjohnso commented:
I read through the notes (sorry I missed the meeting) and added a few comments.

One thing to also consider is maintenance of the module. If you expose k8s resources directly in the module, you'll need to keep the module updated with api changes. Google will only introduce breaking changes during an API version bump, but they will add new, non-breaking, capabilities to the existing version.

A way around the hand-crafted maintenance would be to auto-generate the ansible module based on the k8s spec. This was feedback I received from the k8s team when I reviewed this module with them internally.  So, another approach to consider would be a contrib script that gets run that auto-generates this entire module.  I'd anticipate that while this sounds clean on the surface, the script would likely have some ugliness to it so you're not exposing resources such as Events or Watches (unless that's what the community would like).

Zempashi commented:
@erjohnso ok, thanks for pointing at this. I'll need to figure out why I get a 401 statuscode

detiber commented:
@erjohnso Indeed, the goal was for the client library to be a temporary hack until a point when kubernetes exposes a go client that can be exported as a shared library and at that point the python wrapper would simply become a python interface for the go client.

The autogeneration based on the spec for the python client is interesting though, and is definitely something that could be done until a proper go client is available (and exposable as a shared library).

gregdek commented:
So I think @detiber and @erjohnso have the ball on working this one out.  I regard the two of you as the experts and eventual maintainers. If/when one/both of you think that it's actually good enough to ship, mark it "shipit" and we will ship it.

detiber commented:
@erjohnso I've actually been working on a [project](https://github.com/ansible-swagger-generator/ansible-swagger-generator) to convert a swagger spec into a set of Ansible modules in the hope that it'll provide us with a nice and maintainable way to create maintainable modules for both Kubernetes and OpenShift. Hopefully within the next week or two I will have something that will be able to generate some rough module code.

There will definitely still be a need for a more generic Kubernetes module that can just take a valid json doc and deploy it to a cluster. To use the AWS modules as an example, this one would be the cloudformation module compared to the individual AWS service modules.

detiber commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,449 @@
+
+    if state in ['replace', 'put']:
+        changed, body = k8s_replace_resource(module, url, data)
+        module.exit_json(changed=changed, api_response=body)
```
Instead of a 'replace' state I'd rather see an additional module parameter used with present.

detiber commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,449 @@
+        changed, body = k8s_replace_resource(module, url, data)
+        module.exit_json(changed=changed, api_response=body)
+
+    if state in ['update', 'patch']:
```
Updates should be done when state == 'present' and the current config in the cluster does not match.

detiber commented on clustering/kubernetes.py:
```diff
@@ -0,0 +1,449 @@
+                                     "/api/v1/namespaces/kube-system"
+                                     "/services?labelSelector=kubernetes"
+                                     ".io/cluster-service=true")
+            module.exit_json(changed=False, api_response=body)
```
I would expect these to exist in a 'facts' module instead of being part of this module.

pradeepto commented:
Any idea when this module will be merged? Is there any blocker?

gregdek commented:
Hi @detiber and @erjohnso. Thanks to both of you for working on this one.

Putting this in needs_revision for now. The basic complaint from the core team on review is that it doesn't add any significant value over invoking the kube api/cli directly. From discussion about what we should require from a new module:

"must have added value on top of existing tool, checkmode/change management/detection, options instead of passing existing configs"

So we should consider if we can make this module do some of those things.

erjohnso commented:
I'm in favor of scuttling this and closing the PR. I'd prefer to see something more in-line with what the community wants.

If there are no objections, please feel free to close.

jimi-c commented:
Hi all, I've taken a look at this module and have made some revisions based on testing/best practices. My working branch can be found here, if anyone wants to take a look: https://github.com/ansible/ansible-modules-extras/compare/erjohnso-google-kubernetes

Most notably, my changes in 62fdb2bbe35ef8e1514ba50f13014026fa42fe15 are:
1. Removed kubectl functionality. We'll move that into a different module in the future. Also removed post/put/patch/delete options, as they are not Ansible best practice.
2. Added an 'insecure' option and made the password param optional, to enable the use of the local insecure port.
3. Allowed the data (both inline and from the file) to support multiple items via a list. This is common in YAML files where mutliple docs are used to create/remove multiple resources in one shot.

We're currently doing some more internal testing with this, but if anyone else would like to review it please do so and let us know if you find any problems with it.

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

erjohnso commented:
@jimi-c - your changes look great. If you're all set with it, are we ready to pull these changes in and merge?

jimi-c commented:
@erjohnso thanks! Merged now.

icereed commented:
Hi @detiber,
Is there any progress on the ansible module you mentioned above? 
Would be very interested to use it. :)
